<launch>
  <arg name="input/rois_number" default="6"/>
  <arg name="input/rois0" default="rois0"/>
  <arg name="input/camera_info0" default="/camera_info0"/>
  <arg name="input/rois1" default="rois1"/>
  <arg name="input/camera_info1" default="/camera_info1"/>
  <arg name="input/rois2" default="rois2"/>
  <arg name="input/camera_info2" default="/camera_info2"/>
  <arg name="input/rois3" default="rois3"/>
  <arg name="input/camera_info3" default="/camera_info3"/>
  <arg name="input/rois4" default="rois4"/>
  <arg name="input/camera_info4" default="/camera_info4"/>
  <arg name="input/rois5" default="rois5"/>
  <arg name="input/camera_info5" default="/camera_info5"/>
  <arg name="input/rois6" default="rois6"/>
  <arg name="input/camera_info6" default="/camera_info6"/>
  <arg name="input/rois7" default="rois7"/>
  <arg name="input/camera_info7" default="/camera_info7"/>
  <arg name="input/pointcloud" default="/sensing/lidar/top/rectified/pointcloud"/>
  <arg name="output/objects" default="objects"/>
  <arg name="data_path" default="$(env HOME)/autoware_data" description="packages data and artifacts directory path"/>
  <arg name="model_name" default="pointpainting" description="options: `pointpainting`"/>
  <arg name="model_path" default="$(var data_path)/image_projection_based_fusion"/>
  <arg name="model_param_path" default="$(find-pkg-share autoware_image_projection_based_fusion)/config/pointpainting.param.yaml"/>
  <arg name="ml_package_param_path" default="$(var model_path)/$(var model_name)_ml_package.param.yaml"/>
  <arg name="class_remapper_param_path" default="$(var model_path)/detection_class_remapper.param.yaml"/>
  <arg name="sync_param_path" default="$(find-pkg-share autoware_image_projection_based_fusion)/config/fusion_common.param.yaml"/>
  <arg name="build_only" default="false" description="shutdown node after TensorRT engine file is built"/>

  <arg name="use_pointcloud_container" default="false" description="use pointcloud_container"/>
  <arg name="pointcloud_container_name" default="pointcloud_container" description="pointcloud_container name"/>

  <!-- for eval variable-->
  <arg name="input_rois_number" default="$(var input/rois_number)"/>

  <!-- debug -->
  <arg name="input/image0" default="/image_raw0"/>
  <arg name="input/image1" default="/image_raw1"/>
  <arg name="input/image2" default="/image_raw2"/>
  <arg name="input/image3" default="/image_raw3"/>
  <arg name="input/image4" default="/image_raw4"/>
  <arg name="input/image5" default="/image_raw5"/>
  <arg name="input/image6" default="/image_raw6"/>
  <arg name="input/image7" default="/image_raw7"/>

  <group if="$(var use_pointcloud_container)">
    <load_composable_node target="$(var pointcloud_container_name)">
      <composable_node pkg="autoware_image_projection_based_fusion" plugin="autoware::image_projection_based_fusion::PointPaintingFusionNode" name="pointpainting">
        <remap from="~/input/pointcloud" to="$(var input/pointcloud)"/>
        <remap from="~/output/objects" to="$(var output/objects)"/>
        <param from="$(var model_param_path)" allow_substs="true"/>
        <param from="$(var ml_package_param_path)" allow_substs="true"/>
        <param from="$(var sync_param_path)"/>
        <param from="$(var class_remapper_param_path)"/>
        <param name="rois_number" value="$(var input/rois_number)"/>

        <!-- This parameter shall NOT be included in param file. See also: https://github.com/autowarefoundation/autoware_universe/pull/6169 -->
        <param name="build_only" value="$(var build_only)"/>

        <!-- rois, camera and info -->
        <param name="input/rois0" value="$(var input/rois0)"/>
        <param name="input/camera_info0" value="$(var input/camera_info0)"/>
        <param name="input/image0" value="$(var input/image0)"/>
        <param name="input/rois1" value="$(var input/rois1)"/>
        <param name="input/camera_info1" value="$(var input/camera_info1)"/>
        <param name="input/image1" value="$(var input/image1)"/>
        <param name="input/rois2" value="$(var input/rois2)"/>
        <param name="input/camera_info2" value="$(var input/camera_info2)"/>
        <param name="input/image2" value="$(var input/image2)"/>
        <param name="input/rois3" value="$(var input/rois3)"/>
        <param name="input/camera_info3" value="$(var input/camera_info3)"/>
        <param name="input/image3" value="$(var input/image3)"/>
        <param name="input/rois4" value="$(var input/rois4)"/>
        <param name="input/camera_info4" value="$(var input/camera_info4)"/>
        <param name="input/image4" value="$(var input/image4)"/>
        <param name="input/rois5" value="$(var input/rois5)"/>
        <param name="input/camera_info5" value="$(var input/camera_info5)"/>
        <param name="input/image5" value="$(var input/image5)"/>
        <param name="input/rois6" value="$(var input/rois6)"/>
        <param name="input/camera_info6" value="$(var input/camera_info6)"/>
        <param name="input/image6" value="$(var input/image6)"/>
        <param name="input/rois7" value="$(var input/rois7)"/>
        <param name="input/camera_info7" value="$(var input/camera_info7)"/>
        <param name="input/image7" value="$(var input/image7)"/>
        <extra_arg name="use_intra_process_comms" value="true"/>
      </composable_node>
    </load_composable_node>
  </group>
  <group unless="$(var use_pointcloud_container)">
    <node pkg="autoware_image_projection_based_fusion" exec="pointpainting_fusion_node" name="pointpainting" output="screen">
      <remap from="~/input/pointcloud" to="$(var input/pointcloud)"/>
      <remap from="~/output/objects" to="$(var output/objects)"/>
      <param from="$(var model_param_path)" allow_substs="true"/>
      <param from="$(var ml_package_param_path)" allow_substs="true"/>
      <param from="$(var sync_param_path)"/>
      <param from="$(var class_remapper_param_path)"/>
      <param name="rois_number" value="$(var input/rois_number)"/>

      <!-- This parameter shall NOT be included in param file. See also: https://github.com/autowarefoundation/autoware_universe/pull/6169 -->
      <param name="build_only" value="$(var build_only)"/>

      <!-- rois, camera and info -->
      <param name="input/rois0" value="$(var input/rois0)"/>
      <param name="input/camera_info0" value="$(var input/camera_info0)"/>
      <param name="input/image0" value="$(var input/image0)"/>
      <param name="input/rois1" value="$(var input/rois1)"/>
      <param name="input/camera_info1" value="$(var input/camera_info1)"/>
      <param name="input/image1" value="$(var input/image1)"/>
      <param name="input/rois2" value="$(var input/rois2)"/>
      <param name="input/camera_info2" value="$(var input/camera_info2)"/>
      <param name="input/image2" value="$(var input/image2)"/>
      <param name="input/rois3" value="$(var input/rois3)"/>
      <param name="input/camera_info3" value="$(var input/camera_info3)"/>
      <param name="input/image3" value="$(var input/image3)"/>
      <param name="input/rois4" value="$(var input/rois4)"/>
      <param name="input/camera_info4" value="$(var input/camera_info4)"/>
      <param name="input/image4" value="$(var input/image4)"/>
      <param name="input/rois5" value="$(var input/rois5)"/>
      <param name="input/camera_info5" value="$(var input/camera_info5)"/>
      <param name="input/image5" value="$(var input/image5)"/>
      <param name="input/rois6" value="$(var input/rois6)"/>
      <param name="input/camera_info6" value="$(var input/camera_info6)"/>
      <param name="input/image6" value="$(var input/image6)"/>
      <param name="input/rois7" value="$(var input/rois7)"/>
      <param name="input/camera_info7" value="$(var input/camera_info7)"/>
      <param name="input/image7" value="$(var input/image7)"/>
    </node>
  </group>
</launch>
